clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -5723,16 +5723,14 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (Arg *A = Args.getLastArg(options::OPT_mcmodel_EQ)) {
     StringRef CM = A->getValue();
     bool Ok = false;
-    if (Triple.isOSAIX() && CM == "medium") {
+    if (Triple.isOSAIX() && CM == "medium")
       CM = "large";
-      Ok = true;
-    }
     if (Triple.isAArch64(64)) {
       Ok = CM == "tiny" || CM == "small" || CM == "large";
       if (CM == "large" && RelocationModel != llvm::Reloc::Static)
         D.Diag(diag::err_drv_argument_only_allowed_with)
             << A->getAsString(Args) << "-fno-pic";
-    } else if (Triple.isPPC64()) {
+    } else if (Triple.isPPC64() || Triple.isOSAIX()) {
```

Comment:
AIX seems good now.
However I think we still have one active ppc32 target like SPE working on PPC. So now clang can not accept any code model for target `-target powerpcspe`.
From simple testing, this target supports all these three models.

Reply:
CC @chmeeedalf

Reply:
Hi, may I ask what `-mcmodel=large` does on a 32-bit AIX?

`-mcmodel=large` is generally not supported for 32-bit systems and it seems that AIX does something special here.

GCC's rs6000 doc says:

> Generate PowerPC64 code for the large model: The TOC may be up to 4G in size. Other data and code is only limited by the 64-bit address space.

Reply:
```
int b = 20;
int foo()
{
  return b;
}
```

- Small code model:
```
.foo:
# %bb.0:                                # %entry
	lwz 3, L..C0(2)                         # @b  ;; Use one instruction to load b's address in the TOC. So only 2^16 bytes TOC range can be accessed because of encoding limitation of `lwz` instruction.
	lwz 3, 0(3)
	blr
```

- large code model
```
.foo:
# %bb.0:                                # %entry
	addis 3, L..C0@u(2)
	lwz 3, L..C0@l(3)          ;; Use two instructions to load b's address. So that bigger TOC range can be accessed.
	lwz 3, 0(3)
	blr
```